Associate worker process to build operation id it's working on

eriwen opened the pull request:
### Context
This addresses #1816, but should _not_ be merged until #1818 is handled. The intention is to break down the review into infrastructure/logging changes.

The reason to have a PR here is so we can have reviews on a living branch instead of a set of commits.

[CI Build](https://builds.gradle.org/viewQueued.html?itemId=2770995&tab=queuedBuildOverviewTab)

### Future work
* We still want to avoid the singleton `BuildOperationIdentifierRegistry`, but if I understood @adammurdoch in our conversation earlier this week, we're ok with this for a first pass.

### Contributor Checklist
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew quickCheck <impacted-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs

bmuschko commented on subprojects/base-services/src/main/java/org/gradle/internal/operations/BuildOperationIdentifierRegistry.java:
```diff
@@ -0,0 +1,32 @@
+ */
+package org.gradle.internal.operations;
+
+public class BuildOperationIdentifierRegistry {
```
This class should like be registered as global service and then passed to the relevant classes. For testing purposed I'd probably also introduce an interface. As a result the methods do not need to be static anymore.

bmuschko replied:
Class needs a test.

bmuschko replied:
done

bmuschko commented on subprojects/base-services/src/main/java/org/gradle/internal/operations/BuildOperationIdentifierRegistry.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public static void setCurrentOperationIdentifier(OperationIdentifier operationId) {
+        if (operationId == null) {
```
It might be clearer if we'd introduce another method for removing the operation Id. Maybe something like `clearCurrentOperationIdentifier()`.

bmuschko commented on subprojects/base-services/src/main/java/org/gradle/internal/operations/BuildOperationIdentifierRegistry.java:
```diff
@@ -0,0 +1,32 @@
+public class BuildOperationIdentifierRegistry {
+    private static ThreadLocal<OperationIdentifier> localOperationId = new ThreadLocal<OperationIdentifier>();
+
+    public static OperationIdentifier getCurrentOperationIdentifier() {
```
Likely needs a `@Nullable` annotation.

bmuschko replied:
done

bmuschko commented on subprojects/base-services/src/main/java/org/gradle/internal/operations/OperationIdentifier.java:
```diff
@@ -0,0 +1,51 @@
+ */
+package org.gradle.internal.operations;
+
+public class OperationIdentifier {
```
As silly as it might sounds but let's make sure instances can be compared by adding a test e.g. similar to `DefaultModuleComponentIdentifierTest`.

bmuschko commented on subprojects/logging/src/main/java/org/gradle/internal/logging/events/RenderableOutputEvent.java:
```diff
@@ -16,13 +16,17 @@
+    private OperationIdentifier buildOperationIdentifier;
 
-    protected RenderableOutputEvent(long timestamp, String category, LogLevel logLevel) {
+    protected RenderableOutputEvent(long timestamp, String category, LogLevel logLevel, @Nullable org.gradle.internal.operations.OperationIdentifier buildOperationIdentifier) {
```
No need to have fully-qualified class as we already imported class.

eriwen commented:
@ghale requesting your optional review. You had previously reviewed https://github.com/gradle/gradle/pull/1838. I took over the task while @lacasseio is on vacation. 

I think I addressed all of your comments on that PR.

bmuschko commented on subprojects/logging/src/main/java/org/gradle/internal/logging/progress/DefaultProgressLoggerFactory.java:
```diff
@@ -40,25 +42,40 @@ public ProgressLogger newOperation(Class loggerCategory) {
+        // Assign fake operation ID so that we can still complete logging headers correctly.
+        // This will go away as we force all progress loggers to provide a build operation id
+        if (operationIdentifier == null) {
+            // Decrement from -1 to avoid any conflict between this ID and build operation IDs
```
I don't quite understand the problem we are solving here. Maybe we should talk about it.

eriwen replied:
I never want a "build operation" OperationId to _ever_ be equal to a progress logger Operation id, because I'd hate for log grouping to be messed up because somehow a fake operation id for progress logging had the same ID as a real build operation. Make sense?

bmuschko replied:
Shouldn't we have a separate field to represent that instead?

bmuschko commented on subprojects/logging/src/main/java/org/gradle/internal/logging/progress/ProgressLoggerFactory.java:
```diff
@@ -36,5 +38,16 @@
+     * with the given build operation id.
+     *
+     * @param loggerCategory The logger category.
+     * @param buildOperationId OperationIdentifier to be associated with all logs produced by
```
I am not sure if the Javadoc would render properly if the parameter description is on two lines.

bmuschko commented on subprojects/logging/src/main/java/org/gradle/internal/logging/progress/ProgressLoggerFactory.java:
```diff
@@ -36,5 +38,16 @@
+     */
+    ProgressLogger newOperation(Class<?> loggerCategory, OperationIdentifier buildOperationId);
+
     ProgressLogger newOperation(Class<?> loggerClass, ProgressLogger parent);
```
Could we have the use case where we need the `parent` and `OperationIdentifier`?

eriwen replied:
Not yet, I don't think. `TaskExecutionLogger` is the only thing that does this, and I expect we'll change up how it does progress logging to build operations before too long.

eriwen replied:
Done in 71a080f

bmuschko commented on subprojects/base-services/src/test/groovy/org/gradle/internal/operations/BuildOperationIdentifierRegistryTest.groovy:
```diff
@@ -0,0 +1,21 @@
+import spock.lang.Specification
+import spock.lang.Subject
+
+@Subject(BuildOperationIdentifierRegistry)
```
I assume we are just adding this for information purposes?

eriwen replied:
Yeah, as with other tests I did this. I'm sure the IDE finds it anyway, but this makes it _super_ clear. I don't think we have a convention here.

bmuschko commented on subprojects/base-services/src/test/groovy/org/gradle/internal/operations/BuildOperationIdentifierRegistryTest.groovy:
```diff
@@ -0,0 +1,21 @@
+
+@Subject(BuildOperationIdentifierRegistry)
+class BuildOperationIdentifierRegistryTest extends Specification {
+    def "able to set and retrieve current operation across threads"() {
```
should have `expect: BuildOperationIdentifierRegistry.getCurrentOperationIdentifier() == null` to verify the initial state.

eriwen replied:
Done

eriwen commented on subprojects/base-services/src/main/java/org/gradle/internal/progress/BuildOperationState.java:
```diff
@@ -17,7 +17,7 @@
 
 public interface BuildOperationState {
-    Object getId();
+    OperationIdentifier getId();
```
@jjohannes I would like your awareness of this change. 

I moved `OperationIdentifier` into `base-services` package `org.gradle.internal.progress`. This fixes one of the problematic dependencies with ProgressLogging. 

I decided it best to introduce a strong type here now that the class is available so that I could manage this as an `OperationIdentifier` instead of a plain `Object` in `DefaultBuildOperationExecutor`. 

Please acknowledge, and optionally, provide feedback on the parts of this PR that touch Build Operations and Workers infrastructure.

eriwen replied:
@jjohannes nevermind, I'm backing out this change for now because it is not strictly necessary and I found a few more places I'd have to use this type.

adammurdoch commented on subprojects/base-services/src/main/java/org/gradle/internal/progress/BuildOperationDescriptor.java:
```diff
@@ -23,14 +23,14 @@
 public class BuildOperationDescriptor {
-    private final Object id;
-    private final Object parentId;
+    private final OperationIdentifier id;
```
We don't necessarily need to strongly type the build operation ids to make logging work well. I would do that as a separate move.

adammurdoch commented on subprojects/core/src/main/java/org/gradle/process/internal/worker/request/WorkerAction.java:
```diff
@@ -64,25 +66,27 @@ public void execute(WorkerProcessContext workerProcessContext) {
         }
         try {
             Method method = workerImplementation.getDeclaredMethod(methodName, paramTypes);
+            BuildOperationIdentifierRegistry.setCurrentOperationIdentifier(operationIdentifier);
```
Need to clear the build operation id after invoking the method.

adammurdoch commented on subprojects/core/src/main/java/org/gradle/internal/progress/DefaultBuildOperationExecutor.java:
```diff
@@ -160,7 +160,7 @@ public void run(RunnableBuildOperation buildOperation) {
         DefaultBuildOperationState operationBefore = this.currentOperation.get();
         this.currentOperation.set(currentOperation);
-
+        BuildOperationIdentifierRegistry.setCurrentOperationIdentifier((OperationIdentifier) this.currentOperation.get().getId());
```
should restore the parent's build operation id at completion of this operation

adammurdoch commented on subprojects/logging/src/main/java/org/gradle/internal/logging/console/ProgressOperation.java:
```diff
@@ -16,7 +16,7 @@
 package org.gradle.internal.logging.console;
 
-import org.gradle.internal.logging.events.OperationIdentifier;
+import org.gradle.internal.progress.OperationIdentifier;
```
Moving `OperationIdentifier` around isn't really necessary for this work. We should do that as a separate move.

adammurdoch commented on subprojects/logging/src/main/java/org/gradle/internal/logging/events/LogEvent.java:
```diff
@@ -51,4 +62,34 @@ public void render(StyledTextOutput output) {
         return "[" + getLogLevel() + "] [" + getCategory() + "] " + message;
     }
+
+    public static class Builder {
```
We need to keep garbage to a minimum in the logging infrastructure. I would not use a builder for these events.

adammurdoch commented on subprojects/logging/src/main/java/org/gradle/internal/logging/events/ProgressStartEvent.java:
```diff
@@ -18,28 +18,39 @@
     private final String shortDescription;
     private final String loggingHeader;
     private final String status;
+    private final OperationIdentifier buildOperationId;
```
should be typed `Object`.

adammurdoch commented on subprojects/logging/src/main/java/org/gradle/internal/logging/events/ProgressStartEvent.java:
```diff
@@ -65,7 +76,11 @@ public String toString() {
+        return progressOperationId;
+    }
+
+    public OperationIdentifier getBuildOperationId() {
```
@Nullable?

adammurdoch commented on subprojects/logging/src/main/java/org/gradle/internal/logging/events/RenderableOutputEvent.java:
```diff
@@ -16,13 +16,17 @@
+import org.gradle.internal.progress.OperationIdentifier;
 
 public abstract class RenderableOutputEvent extends CategorisedOutputEvent {
+    private OperationIdentifier buildOperationIdentifier;
```
Should be typed `Object`.

adammurdoch commented on subprojects/logging/src/main/java/org/gradle/internal/logging/events/StyledTextOutputEvent.java:
```diff
@@ -90,6 +63,12 @@ public void render(StyledTextOutput output) {
         }
     }
 
+    public StyledTextOutputEvent.Builder toBuilder() {
```
Again, I would not use a builder here, to avoid garbage.

adammurdoch commented on subprojects/logging/src/main/java/org/gradle/internal/logging/progress/DefaultProgressLoggerFactory.java:
```diff
@@ -40,25 +41,38 @@ public ProgressLogger newOperation(Class loggerCategory) {
         return newOperation(loggerCategory.getName());
     }
 
+    public ProgressLogger newOperation(Class loggerCategory, OperationIdentifier operationIdentifier) {
```
should be typed `Object` and I would call the parameter something like "buildOperationIdentifier" to make it clear whether it refers to a build operation id or a progress operation id.

adammurdoch commented on subprojects/logging/src/main/java/org/gradle/internal/logging/progress/DefaultProgressLoggerFactory.java:
```diff
@@ -40,25 +41,38 @@ public ProgressLogger newOperation(Class loggerCategory) {
+        return init(loggerCategory, null, null);
+    }
+
+    public ProgressLogger newOperation(String loggerCategory, OperationIdentifier operationIdentifier) {
```
Same here: type `Object, maybe rename parameter

adammurdoch commented on subprojects/logging/src/main/java/org/gradle/internal/logging/progress/DefaultProgressLoggerFactory.java:
```diff
@@ -40,25 +41,38 @@ public ProgressLogger newOperation(Class loggerCategory) {
-        return new ProgressLoggerImpl((ProgressLoggerImpl) parentOperation, nextId.getAndIncrement(), loggerCategory, progressListener, timeProvider);
+
+        // Decrement from -1 to avoid any conflict between this ID and build operation IDs
+        OperationIdentifier progressOperationId = new OperationIdentifier(nextId.getAndDecrement());
```
I don't think this is necessary. They're 2 different namespaces

eriwen commented:
Thanks for your feedback @adammurdoch. I believe I have addressed your concerns with 87ac76c4a23a38d2cddbc77b8138c1ed1e252cae.

bmuschko commented on subprojects/logging/src/main/java/org/gradle/internal/logging/console/BuildStatusRenderer.java:
```diff
@@ -82,8 +82,8 @@ public void onOutput(OutputEvent event) {
             // if it has no parent ID, assign this operation as the root operation
-            if (startEvent.getParentId() == null && BUILD_PROGRESS_CATEGORY.equals(startEvent.getCategory())) {
-                rootOperationId = startEvent.getOperationId();
+            if (startEvent.getParentProgressOperationId() == null && BUILD_PROGRESS_CATEGORY.equals(startEvent.getCategory())) {
```
Didn't know that we also have a hard-coded category here. We can clean that up with the next PR that maps build operation descriptor to logging type.

bmuschko commented on subprojects/logging/src/main/java/org/gradle/internal/logging/serializer/ProgressStartEventSerializer.java:
```diff
@@ -25,31 +25,38 @@
         } else {
             encoder.writeBoolean(true);
-            encoder.writeSmallLong(event.getParentId().getId());
+            encoder.writeSmallLong(event.getParentProgressOperationId().getId());
```
Could small long become an issue for bigger builds e.g. 500 subproject that produce a lot of operation IDs?

eriwen replied:
Added a test

bmuschko commented on subprojects/logging/src/test/groovy/org/gradle/internal/logging/serializer/StyledTextOutputEventSerializerTest.groovy:
```diff
@@ -0,0 +1,58 @@
+        result.spans[0].text == "description"
+        result.spans[1].style == StyledTextOutput.Style.Error
+        result.spans[1].text == "error"
+        result.buildOperationId == new OperationIdentifier(42L)
```
Should probably also test the case where operation ID could be null.

eriwen replied:
Test added

eriwen commented:
This has been merged with https://github.com/gradle/gradle/commit/7b84dacf981af34eae48ef02f0cc91bad7b4c6d8

adammurdoch commented on subprojects/base-services/src/main/java/org/gradle/internal/operations/BuildOperationIdentifierRegistry.java:
```diff
@@ -0,0 +1,39 @@
+
+import org.gradle.api.Nullable;
+
+public class BuildOperationIdentifierRegistry {
```
This actually duplicates the work that `DefaultBuildOperationExecutor` does. We should merge them so that `DefaultBuildOperationExecutor` is the single point of reference for what the current operation is.
